kf_lib_data_ingest/test_runner.py

```diff
@@ -0,0 +1,121 @@
+           pipeline runs without modifying data in the target service
+
+    """
+    logging.info('Running standard unit tests for '
+                 f'{ingest_pipeline.ingest_config_dir} ...')
+    # TODO Enable this when dry run mode is implemented
+    # kwargs.update({
+    #     'dry_run': True
+    # })
+    ingest_pipeline.run(TARGET_SERVICE_CONFIG_PATH, **kwargs)
+    logging.info('Standard unit tests passed!')
+
+
+def _run_user_defined_tests(ingest_pipeline, restricted_urls=None):
+    """
+    Runs user defined unit tests for the ingest package
+
+    Test Requirements
+    -----------------
+    User defined tests must be placed in the `tests` directory inside of the
+    ingest package dir (ingest_pipeline.ingest_config_dir).
+
+    Since these tests run after the standard tests complete, they are meant
+    to operate on the ingestion outputs of the standard test.
+
+    pytest is used to run these tests and so all tests should conform to the
+    pytest standard.
+
+    Protection Against User Error
+    -----------------------------
+    Because these tests are user defined, there is potential for accidental
+    loading of data into target services. It seems highly unlikely that the
+    user would do this, but it could happen.
+
+    To prevent against this, the test runner will mock out any POST/PATCH http
+    requests to the `restricted_urls` and raise an exception if
+    any attempts are made.
+
+    :param ingest_pipeline: the ingest pipeline containing a reference to the
+    ingest package dir for which user defined tests will run
+    :type ingest_pipeline: DataIngestPipeline
+    :param restricted_urls: list of restricted target service URLs
+    :type: list
+    """
+    logging.info('Check if user defined unit tests exist for '
+                 f'{ingest_pipeline.ingest_config_dir}')
+
+    user_defined_test_dir = os.path.join(ingest_pipeline.ingest_config_dir,
+                                         'tests')
+    # No user defined tests exist
+    if not os.path.isdir(user_defined_test_dir):
+        return
+
+    # Execute unit tests
+    logging.info('Running user defined unit tests '
+                 f'{user_defined_test_dir} ...')
+
+    # Mock post/patch requests to restricted target service URLs
+    # Return exception on attempted requests
+    if not restricted_urls:
+        restricted_urls = DEFAULT_RESTRICTED_TARGET_URLS
+
+    with requests_mock.Mocker(real_http=True) as m:
+        msg = ('No POST or PATCH requests allowed to restricted '
+               f'target services in user defined unit tests! Check to make '
+               'sure your unit tests are not attempting to load data into '
+               f'target services at any of these URLs: {restricted_urls}')
+        for method_name in ['post', 'patch']:
+            for url in restricted_urls:
+                m.register_uri(method_name.upper(), url, exc=Exception(msg))
+
+        exit_code = pytest.main([user_defined_test_dir])
```

Comment:
pytest output currently goes to stdout. Should we try to add it to our ingest log?

Reply:
It could go in its own file